## Project: Poverty

In [2]:
import pandas as pd

### Poverty data fetched from World Bank
The data includes 108 columns including general information like country, year, reporting level, survey comparability, etc., and other digital data.

Data Source: World Bank`s Poverty and Inequality Platform (PIP), extracted by owid. https://github.com/owid/poverty-data

In [3]:
df = pd.read_csv("poverty_dataset.csv")
df.head()

country  year reporting_level welfare_type  ppp_version  survey_year  \
0  Albania  1996        national  consumption         2011       1996.0   
1  Albania  2002        national  consumption         2011       2002.0   
2  Albania  2005        national  consumption         2011       2005.0   
3  Albania  2008        national  consumption         2011       2008.0   
4  Albania  2012        national  consumption         2011       2012.0   

   survey_comparability  headcount_ratio_international_povline  \
0                   0.0                               0.920669   
1                   1.0                               1.570843   
2                   1.0                               0.860527   
3                   1.0                               0.313650   
4                   1.0                               0.849754   

   headcount_ratio_lower_mid_income_povline  \
0                                 11.174149   
1                                 14.132118   
2                                  8.715685   
3                                  5.250542   
4                                  6.182414   

   headcount_ratio_upper_mid_income_povline  ...  decile8_thr  decile9_thr  \
0                                 44.618417  ...         8.85        10.92   
1                                 49.669635  ...         8.83        11.58   
2                                 38.545254  ...        10.02        12.78   
3                                 31.110345  ...        10.74        13.62   
4                                 34.528906  ...        10.52        13.26   

       gini       mld  polarization  palma_ratio  s80_s20_ratio  \
0  0.270103  0.119104      0.241293     0.928335       3.945872   
1  0.317390  0.164812      0.268982     1.215056       4.831625   
2  0.305957  0.154413      0.254529     1.142718       4.662236   
3  0.299847  0.148893      0.247311     1.114657       4.395911   
4  0.289605  0.138417      0.249988     1.041193       4.272573   

   p90_p10_ratio  p90_p50_ratio  p50_p10_ratio  
0       3.568627       1.889273       1.888889  
1       3.979381       2.090253       1.903780  
2       3.872727       1.978328       1.957576  
3       3.574803       1.956897       1.826772  
4       3.632877       1.941435       1.871233  

[5 rows x 108 columns]

#### headcount data
Columns 8 to 29 are the headcount data. They basiclly tell the percentage or the number of people living under certain line (for instance, certain expenditure per day per person or certain percentage of the population).

Below is an example of column 8: % of population living in households with an income or expenditure per person below the International Poverty Line: 1.90 dollar a day (in 2011 prices) or 2.15 dollar a day (in 2017 prices).

In [5]:
df["headcount_ratio_international_povline"].describe()

count    4877.000000
mean       11.081565
std        18.190433
min         0.000000
25%         0.278387
50%         2.041412
75%        13.390963
max        96.871427
Name: headcount_ratio_international_povline, dtype: float64

#### shortfall and gap data
Columns 30 to 62 are the shortfall data.They describe the shortfall to the line that ensures the basic needs for living.

Columns 63 to 70 are the gap data. It is defined as the mean shortfall of the total population from the poverty line counting the non-poor as having zero shortfall and expressed as a percentage of the poverty line.

Below shows the total shortfall from the upper-middle income poverty line: 5.50 dollar a day (in 2011 prices) or 6.85 dollar a day (in 2017 prices). This is the amount of money that would be theoretically needed to lift the incomes or expenditure of all people in poverty up to the poverty line. 

However this is not a measure of the actual cost of eliminating poverty, since it does not take into account the costs involved in making the necessary transfers nor any changes in behaviour they would bring about.

In [15]:
df["avg_shortfall_upper_mid_income_povline"].sort_values()

2155    0.059127
820     0.219790
255     0.227310
2156    0.240166
734     0.280244
          ...   
2525         NaN
3809         NaN
3810         NaN
3812         NaN
4712         NaN
Name: avg_shortfall_upper_mid_income_povline, Length: 4877, dtype: float64

As you can see, there are many blank cells in the dataset. It would be dealt later in the data processing part.

#### mean and threshold data for different deciles of people
Columns 71 to 100 are the mean and threshold data. They describe the average income or the income threshold for certain decile (tenth of the population).

For instance, column "decile10_share" tells the income or expenditure of the richest decile (tenth of the population) as a share of total income or expenditure.

In [16]:
df["decile10_share"].describe()

count    4395.000000
mean       29.434217
std         6.742275
min        16.985789
25%        24.479009
50%        27.494046
75%        32.899965
max        61.491135
Name: decile10_share, dtype: float64

#### other data or index related
Columns 101 to 108 are the data and indexs related to poverty. Such as Gini index or MLD.

In [17]:
df["gini"].describe()

count    4401.000000
mean        0.375645
std         0.088840
min         0.177920
25%         0.308719
50%         0.355622
75%         0.427676
max         0.657556
Name: gini, dtype: float64

### Tree data
The first visualization would be a tree of countries divided by geographical location. So this section would focus on the transformation from raw data to a tree.

In [19]:
new_df = pd.DataFrame({"country": df["country"], "year": df["year"], "median": df["median"]})
for i in range(len(new_df["country"])):
    if new_df["country"][i] == "Zimbabwe":
        break
    if new_df["country"][i + 1] != new_df["country"][i]:
        continue
    if new_df["country"][i + 1] == new_df["country"][i]:
        new_df = new_df.drop(labels=i)
print(new_df)
new_df.to_excel("tree.xlsx")

        country  year     median
13      Albania  2019  10.831646
16      Algeria  2011   7.158526
19       Angola  2018   1.901955
51    Argentina  2020  12.176018
73      Armenia  2020   5.800005
...         ...   ...        ...
4872     Zambia  2010   1.353084
4873     Zambia  2015   1.570981
4874   Zimbabwe  2011   3.798868
4875   Zimbabwe  2017   2.874558
4876   Zimbabwe  2019   2.654817

[2617 rows x 3 columns]


For the generation of tree data, we need extra data of countries divided by regions. 

Source: United Nations, extracted by Curran Kelleher. https://github.com/curran/data/tree/gh-pages/un/placeHierarchy

In [20]:
from anytree.importer import JsonImporter
from anytree.exporter import JsonExporter
from anytree import RenderTree, LevelOrderIter, Node
import json

importer = JsonImporter()
exporter = JsonExporter(indent=2, sort_keys=True)
country_js = open("countries.json")
country_data = json.load(country_js)
root = importer.import_(json.dumps(country_data))  # convert json to str
print(RenderTree(root))

AnyNode(data={'id': 'World'})
├── AnyNode(data={'id': 'Asia'})
│   ├── AnyNode(data={'id': 'Southern Asia'})
│   │   ├── AnyNode(data={'id': 'Afghanistan'})
│   │   ├── AnyNode(data={'id': 'Bangladesh'})
│   │   ├── AnyNode(data={'id': 'Bhutan'})
│   │   ├── AnyNode(data={'id': 'India'})
│   │   ├── AnyNode(data={'id': 'Iran'})
│   │   ├── AnyNode(data={'id': 'Maldives'})
│   │   ├── AnyNode(data={'id': 'Nepal'})
│   │   ├── AnyNode(data={'id': 'Pakistan'})
│   │   └── AnyNode(data={'id': 'Sri Lanka'})
│   ├── AnyNode(data={'id': 'Western Asia'})
│   │   ├── AnyNode(data={'id': 'Armenia'})
│   │   ├── AnyNode(data={'id': 'Azerbaijan'})
│   │   ├── AnyNode(data={'id': 'Bahrain'})
│   │   ├── AnyNode(data={'id': 'Cyprus'})
│   │   ├── AnyNode(data={'id': 'Georgia'})
│   │   ├── AnyNode(data={'id': 'Iraq'})
│   │   ├── AnyNode(data={'id': 'Israel'})
│   │   ├── AnyNode(data={'id': 'Jordan'})
│   │   ├── AnyNode(data={'id': 'Kuwait'})
│   │   ├── AnyNode(data={'id': 'Lebanon'})
│   │   ├──

To give audience a direct impact, we add the median data (The level of income or expenditure per day below which half of the population live.) to each country. In order to achieve that, we need to match 2 datasets, while the matching point is the country name.

In [21]:
df_tree = pd.read_excel("out.xlsx")
for i in range(len(df_tree["country"])):
    for node in LevelOrderIter(root):
        if node.data["id"] == df_tree["country"][i]:
            node.median = df_tree["median"][i]

for node in LevelOrderIter(root):
    node.name = node.data["id"]
    del node.data
print(RenderTree(root))

AnyNode(median=6.92, name='World')
├── AnyNode(name='Asia')
│   ├── AnyNode(name='Southern Asia')
│   │   ├── AnyNode(name='Afghanistan')
│   │   ├── AnyNode(median=3.110258264324849, name='Bangladesh')
│   │   ├── AnyNode(median=6.702652894737087, name='Bhutan')
│   │   ├── AnyNode(median=4.103054962429613, name='India')
│   │   ├── AnyNode(median=10.22163073544963, name='Iran')
│   │   ├── AnyNode(median=14.35191939823969, name='Maldives')
│   │   ├── AnyNode(median=3.16524183029679, name='Nepal')
│   │   ├── AnyNode(median=3.788850780905248, name='Pakistan')
│   │   └── AnyNode(median=6.170867326898343, name='Sri Lanka')
│   ├── AnyNode(name='Western Asia')
│   │   ├── AnyNode(median=5.800004945187053, name='Armenia')
│   │   ├── AnyNode(median=8.75371479287225, name='Azerbaijan')
│   │   ├── AnyNode(name='Bahrain')
│   │   ├── AnyNode(median=38.98015645283959, name='Cyprus')
│   │   ├── AnyNode(median=5.836041408449774, name='Georgia')
│   │   ├── AnyNode(median=5.345247895973078, 

In [ ]:
# then we export the data in a json file for us to use
# with open("tree.json", 'w') as file:
#     exporter.write(root, file)

### Poverty rate
This dataset concludes the poverty rate of each country.

Data Source: OECD. https://data.oecd.org/inequality/poverty-rate.htm

In [23]:
df_poverty_rate = pd.read_csv("poverty_rate.csv")
df_poverty_rate

LOCATION INDICATOR SUBJECT MEASURE FREQUENCY  TIME  Value Flag Codes
0         AUS   POVERTY    0_17      RT         A  2012  0.129        NaN
1         AUS   POVERTY    0_17      RT         A  2014  0.130        NaN
2         AUS   POVERTY    0_17      RT         A  2016  0.125        NaN
3         AUS   POVERTY    0_17      RT         A  2018  0.133        NaN
4         AUS   POVERTY    0_17      RT         A  2020  0.133        NaN
...       ...       ...     ...     ...       ...   ...    ...        ...
2059      ROU   POVERTY  66MORE      RT         A  2015  0.148        NaN
2060      ROU   POVERTY  66MORE      RT         A  2016  0.161        NaN
2061      ROU   POVERTY  66MORE      RT         A  2017  0.185        NaN
2062      ROU   POVERTY  66MORE      RT         A  2018  0.189        NaN
2063      ROU   POVERTY  66MORE      RT         A  2019  0.199        NaN

[2064 rows x 8 columns]

It has been divided in 4 age categories: 0-17, 18-65, >65 and all.

In [27]:
df_poverty_rate["SUBJECT"].unique()

array(['0_17', 'TOT', '18_65', '66MORE'], dtype=object)

In [28]:
df_poverty_rate["Value"].describe()

count    2064.000000
mean        0.125180
std         0.064124
min         0.018000
25%         0.082000
50%         0.115000
75%         0.156000
max         0.478000
Name: Value, dtype: float64

### Further data engineering
#### Imputation of missing values
As we talked before, there are many blank cells in the dataset. One way of dealing it is to delete the column. However, due to the frequent miss of data from poverty-striken countries, it is almost impossible to continue working if we delete the column every time. So I choose imputation.

One naive way is to use univariate imputation. However it does not seem to fit the data processing of this massive dataset. Thus, I shall use multivariate imputation and I think KNN method is suitable in this case.

KNN method fills in missing values using the k-Nearest Neighbors approach. By default, a euclidean distance metric that supports missing values is used to find the nearest neighbors. Each missing feature is imputed using values from nearest neighbors that have a value for the feature. The feature of the neighbors are averaged uniformly or weighted by distance to each neighbor.

In [22]:
import numpy as np
from sklearn.impute import KNNImputer

X = df.drop(
    ["country", "year", "reporting_level", "welfare_type", "ppp_version", "survey_year", "survey_comparability"],
    axis=1)
imp = KNNImputer(n_neighbors=2, weights="uniform")
imputed_X = np.round_(imp.fit_transform(X), decimals=3)
print(imputed_X)

[[ 0.921 11.174 44.618 ...  3.569  1.889  1.889]
 [ 1.571 14.132 49.67  ...  3.979  2.09   1.904]
 [ 0.861  8.716 38.545 ...  3.873  1.978  1.958]
 ...
 [21.58  47.857 77.945 ...  6.713  2.774  2.42 ]
 [34.206 61.584 84.109 ...  6.575  3.07   2.142]
 [39.755 64.51  85.016 ...  8.191  3.4    2.409]]


Now we have a workable matrix which is also suitable for further dimention reduction.